go-controller/pkg/ovn/controller/egress_services/egress_services_controller.go

```diff
@@ -0,0 +1,365 @@
+
+	klog.Infof("Starting Egress Services Controller")
+
+	if !cache.WaitForNamedCacheSync("egressservices", c.stopCh, c.servicesSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		klog.Infof("Synchronization failed")
+		return
+	}
+
+	if !cache.WaitForNamedCacheSync("egressserviceendpointslices", c.stopCh, c.endpointSlicesSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		klog.Infof("Synchronization failed")
+		return
+	}
+
+	if !cache.WaitForNamedCacheSync("egressservicenodes", c.stopCh, c.nodesSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		klog.Infof("Synchronization failed")
+		return
+	}
+
+	klog.Infof("Repairing Egress Services")
+	err := c.repair()
+	if err != nil {
+		klog.Errorf("Failed to repair Egress Services entries: %v", err)
+	}
+
+	err = c.initClusterEgressPolicies(c.nbClient)
+	if err != nil {
+		klog.Errorf("Failed to init Egress Services cluster policies: %v", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runServiceWorker(wg)
+			}, time.Second, c.stopCh)
+		}()
+	}
+
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runNodeWorker(wg)
+			}, time.Second, c.stopCh)
+		}()
+	}
+
+	go c.checkNodesReachability()
+
+	// wait until we're told to stop
+	<-c.stopCh
+
+	klog.Infof("Shutting down Egress Services controller")
+	c.servicesQueue.ShutDown()
+	c.nodesQueue.ShutDown()
+
+	wg.Wait()
+}
+
+// This takes care of syncing stale data which we might have in OVN if
+// there's no ovnkube-master running for a while.
+// It deletes all logical router policies from OVN that belong to services which are no longer
+// egress services, and the policies of endpoints that do not belong to an egress service.
+// In addition, it removes the egress service labels of deleted services from the nodes.
+func (c *Controller) repair() error {
+	c.Lock()
+	defer c.Unlock()
+
+	svcKeyToAllV4Endpoints := map[string]sets.String{}
+	svcKeyToAllV6Endpoints := map[string]sets.String{}
+	svcKeyToConfiguredV4Endpoints := map[string][]string{}
+	svcKeyToConfiguredV6Endpoints := map[string][]string{}
+	services, _ := c.serviceLister.List(labels.Everything())
+
+	for _, svc := range services {
+		if util.HasEgressSVCAnnotation(svc) && util.HasEgressSVCHostAnnotation(svc) &&
+			util.ServiceTypeHasLoadBalancer(svc) && len(svc.Status.LoadBalancer.Ingress) > 0 {
+			var err error
+			key, _ := cache.MetaNamespaceKeyFunc(svc)
+			conf, err := util.ParseEgressSVCAnnotation(svc)
+			if err != nil && !util.IsAnnotationNotSetError(err) {
+				klog.Errorf("can't parse %s egress service configuration, err: %v", key, err)
+				continue
+			}
+			svcHost, _ := util.GetEgressSVCHost(svc)
+			nodeState, ok := c.nodes[svcHost]
+			if !ok {
+				nodeState, err = c.nodeStateFor(svcHost)
```

Comment:
I added another check that the node object still exists, ready and matches the selectors (which I missed). this pretty much matches what we do when we select a node for a service.
we can assume a node isn't "draining" here (like we do when we choose it for the first time) because it is ready, if it fails the reachability check later its services will be reallocated, and (in theory :sweat_smile:) we shouldn't miss removing an lrp that we should have.